File: leet-code/316. Remove Duplicate Letters/RemoveDuplicateLetters316.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;

public class RemoveDuplicateLetters316 {
    public String removeDuplicateLetters(String s) {
        Map<Character, Integer> counter = new HashMap<>();
        Map<Character, Boolean> complete = new HashMap<>();
        Deque<Character> stack = new ArrayDeque<>();

        for (char c : s.toCharArray()) {
            if (counter.containsKey(c)) {
                counter.put(c, counter.get(c) + 1);
            } else {
                counter.put(c, 1);
            }
        }

        for (char c : s.toCharArray()) {
            counter.put(c, counter.get(c) - 1);

            if (complete.containsKey(c) && complete.get(c)) {
                continue;
            }

            while (!stack.isEmpty() && stack.peek() > c && counter.get(stack.peek()) > 0) {
                complete.put(stack.pop(), false);
            }

            complete.put(c, true);
            stack.push(c);
        }

        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.append(stack.pollLast());
        }
        return sb.toString();
    }
}

File: leet-code/316. Remove Duplicate Letters/RemoveDuplicateLetters316.kt
import java.util.*

class RemoveDuplicateLetters316_2 {
    fun removeDuplicateLetters(s: String): String {
        val counter: MutableMap<Char, Int> = mutableMapOf()
        val complete: MutableMap<Char, Boolean> = mutableMapOf()
        val stack: Deque<Char> = LinkedList()

        for (c in s) {
            counter[c] = counter.getOrDefault(c, 0) + 1
        }

        for (c in s) {
            counter[c] = counter[c]!! - 1

            if (complete.containsKey(c) && complete[c]!!) {
                continue
            }

            while (!stack.isEmpty() && stack.peek() > c && counter[stack.peek()]!! > 0) {
                complete[stack.pop()] = false
            }

            complete[c] = true
            stack.push(c)
        }

        val sb: StringBuilder = StringBuilder()
        while (!stack.isEmpty()) {
            sb.append(stack.pollLast())
        }
        return sb.toString()
    }
}